tools/perf_util/src/main/java/com/msft/send_perf_metrics/App.java

```diff
@@ -94,63 +90,86 @@ private static void processPerfMetrics(final List<Path> perf_metrics, String com
-			st.setString(++i, (String) json_object.get("Optimizer"));
-			st.setInt(++i, (int)(long) json_object.get("BatchSize"));
-
-			// non-key section
-			JSONObject properties = (JSONObject) json_object.get("DerivedProperties");
-			if (properties != null) {
-				if (properties.get("SeqLen") == null) 				//  mysql allows null value in unique key column
-					st.setNull(++i, Types.INTEGER);
-				else
-					st.setInt(++i, Integer.parseInt((String) properties.get("SeqLen")));
-
-				if (properties.get("PredictionsPerSeq") == null) 				//  mysql allows null value in unique key column
-					st.setNull(++i, Types.INTEGER);
-				else
-					st.setInt(++i, Integer.parseInt((String) properties.get("PredictionsPerSeq")));
+		// field name -> json value
+		Map<String, Object> field_mapping = new HashMap();
+		Set<String> update_on_duplicate_fields =
+			new HashSet<> (Arrays.asList("AvgTimePerBatch", "Throughput", "StabilizedThroughput", "TotalTime", "AvgCPU", "Memory"));
+
+		field_mapping.put("BatchId", batch_id);
+		field_mapping.put("CommitId", commit_id.substring(0, 8));
+		json_object.forEach((key, value) -> {
+			if (key.equals("DerivedProperties")) {
+				JSONObject properties = (JSONObject) json_object.get("DerivedProperties");
+				properties.forEach((sub_key, sub_value) -> {
+					field_mapping.put((String)sub_key, sub_value);
+				});
+			} else if (key.equals("Memory")) {
+				field_mapping.put("Memory", (long) value >> 20);
 			} else {
-				st.setNull(++i, Types.INTEGER);
-				st.setNull(++i, Types.INTEGER);
+				field_mapping.put((String)key, value);
 			}
+		});
 
-			st.setInt(++i, (int)(long)  json_object.get("NumOfBatches"));
-			st.setInt(++i, (int)(long)  json_object.get("WeightUpdateSteps"));
-			st.setInt(++i, (int)(long)  json_object.get("Round"));
-			st.setInt(++i, (int)(long)  json_object.get("GradAccSteps"));
-			st.setFloat(++i, (float)(double) json_object.get("AvgTimePerBatch"));  // ms
-			st.setFloat(++i, (float)(double) json_object.get("Throughput"));  // examples/sec
-			st.setFloat(++i, (float)(double) json_object.get("StabilizedThroughput"));  // examples/sec
-			st.setFloat(++i, (float)(double) json_object.get("TotalTime"));  // secs
-			st.setInt(++i, (int)(long) json_object.get("AvgCPU"));
-			st.setInt(++i, (int)((long) json_object.get("Memory") >> 20));  // mb
-			st.setString(++i, (String) json_object.get("RunConfig"));
+		// building sql statement
+		StringBuilder sb = new StringBuilder("INSERT INTO perf_test_training_data (");
+		field_mapping.forEach((key, value) -> {
+			sb.append(key).append(",");
+		});
+		sb.append("Time) values (");
+		for(int i = 0; i < field_mapping.size(); i++) {
+			sb.append("?,");
+		}
+		sb.append("Now()) ON DUPLICATE KEY UPDATE ");
+		update_on_duplicate_fields.forEach((key) -> {
+			if(field_mapping.get(key) != null) {
+				sb.append(key).append("=?,");
+			}
+		});
+
+		//
+		// Map/Set entry order should remain the same as we go thru it
```

Comment:
i'm not very familiar with Java, but a quick look through the docs seems to indicate that there is no guarantee of the HashSet/Map iteration order. How about using LinkedHashSet/Map instead, which preserve the insertion order?

Reply:
if the hashing algo remains the same, then it should, but I'll used linked data structure instead.